accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,191 @@
+
+### Source generator compatibility concerns
+
+Source generators should only generate code which uses public API in framework components they extend.  This public API already needs to meet the strict binary compatibility guarantees as framework API.  As such libraries containing source-generated code will be binary-compatible with future releases, just as if the user wrote this code directly.
+
+Source generators should not generate code which uses public API with "reserved" or "undocumented" functionality in attempts to avoid public API guarantees.  Libraries paired with source generators need to provide the same guarantees as libraries alone do today.
+
+Source generators should avoid emitting public API in a user's projects.  Such API can create a compatibility contract in that user's library which needs to be maintained.  If a source generator produces public API in a user's project by design, that public API must be strictly maintained following the same rules as framework components across versions of the source generator.  Care should be taken to the public API which is generated and should be reviewed by the Framework Design Council.  A similar constraint applies to internal API: users will have source that expects to interact with the generated internal API and the user will expect that code to continue to compile between versions.
+
+Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name collisions with user code and provide affordances for resolving name conflicts.  As much as possible generated code should meet the quality standards of .NET library code, using nullable annotations, meeting framework design guidelines, naming guidelines, etc.  It should feel like a natural extension of the library.
+
+### Deterministic failure
+
+The absence of a required source generator from code that requires it should produce a clear diagnosable error message from the compiler, like the error from a missing reference.  A codebase that was written to make use of a source-generator should never compile and silently omit the source generated code.  For example: if a source generator uses an attribute to activate the generator and then generates extension methods with overloads that are automatically preferred, the source generator should define the attribute in the source generator, not the framework library.  This will ensure that the project fails to compile if the source generator is absent.
```

Comment:
>should define the attribute in the source generator, not the framework library

I think there is an outstanding GitHub issue on this and should be referenced here. This only works if the attribute in question has no inputs (i.e. is just an indication and doesn't contain additional options). The Roslyn source generator will indicate non-existing attributes as errors initially so only the fact that the attribute was emitted can be determined not any of its possible metadata.

Reply:
This document is not meant to call out all possible ways to achieve this behavior, nor link to point in time issues that might make achieving this behavior challenging.
The assertion here is that source-generators when explicitly addressed should be designed in such a way that the compiler will fail if the generator is absent.  
We should avoid the cases where we define framework API specifically for the source generator and then allow code to compile if the source generator is absent.